cirq/qis/states.py

```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
+                with this dimension, i.e., those for which the product of the
+                individual qudit dimensions is equal to the Hilbert space
+                dimension.
+            min_qudit_dimensions: A tuple of integers (n_1, ..., n_k).
+                The associated qid shapes are
+                {(m_1, ..., m_k) : m_i ≥ n_i for all i}.
+        """
+        self.explicit_qid_shapes = explicit_qid_shapes or set()
+        self.unfactorized_total_dimension = unfactorized_total_dimension
+        self.min_qudit_dimensions = min_qudit_dimensions
+
+    def intersection_subset(self, other: '_QidShapeSet'):
+        """Return a subset of the intersection with other qid shape set."""
+        explicit_qid_shapes = self.explicit_qid_shapes & other.explicit_qid_shapes
+        unfactorized_total_dimension = None
+        min_qudit_dimensions = None
+
+        if self.explicit_qid_shapes and other.unfactorized_total_dimension is not None:
+            explicit_qid_shapes |= _intersection_explicit_with_unfactorized_qid_shapes(
+                self.explicit_qid_shapes, other.unfactorized_total_dimension
+            )
+
+        if self.explicit_qid_shapes and other.min_qudit_dimensions:
+            explicit_qid_shapes |= _intersection_explicit_with_min_qudit_dims_qid_shapes(
+                self.explicit_qid_shapes, other.min_qudit_dimensions
+            )
+
+        if self.unfactorized_total_dimension is not None and other.explicit_qid_shapes:
+            explicit_qid_shapes |= _intersection_explicit_with_unfactorized_qid_shapes(
+                other.explicit_qid_shapes, self.unfactorized_total_dimension
+            )
+
+        if self.unfactorized_total_dimension == other.unfactorized_total_dimension:
+            unfactorized_total_dimension = self.unfactorized_total_dimension
+
+        # don't attempt to intersect unfactorized total dimension with
+        # min qudit dimensions
+
+        if self.min_qudit_dimensions is not None and other.explicit_qid_shapes:
+            explicit_qid_shapes |= _intersection_explicit_with_min_qudit_dims_qid_shapes(
+                other.explicit_qid_shapes, self.min_qudit_dimensions
+            )
+
+        # don't attempt to intersect min qudit dimensions with
+        # unfactorized total dimension
+
+        if self.min_qudit_dimensions is not None and other.min_qudit_dimensions is not None:
+            min_qudit_dimensions = _intersection_min_qudit_dims_qid_shapes(
+                self.min_qudit_dimensions, other.min_qudit_dimensions
+            )
+
+        return _QidShapeSet(
+            explicit_qid_shapes=explicit_qid_shapes,
+            unfactorized_total_dimension=unfactorized_total_dimension,
+            min_qudit_dimensions=min_qudit_dimensions,
+        )
+
+    def is_ambiguous(self) -> bool:
+        """Whether the qid shape is ambiguous and cannot be inferred."""
+        if self.min_qudit_dimensions is not None:
+            return True
```

Comment:
return reasons instead of just true/false so people can figure out what's going on

Reply:
Done.